src/app/calendar/components/event-form/event-form.component.ts

```diff
@@ -0,0 +1,149 @@
+    this._id = num;
+    if (this._id < 0) {
+      return;
+    }
+    this.editing = true;
+    this.calendarService.selectEvent(num);
+    this.event$ = this.calendarService.event$;
+    this.fillForm();
+  }
+
+  fillForm(): void {
+    this.event$.pipe(
+      take(1)
+    ).subscribe(res => {
+      this.date?.setValue(moment(res?.date).format("dddd, MMMM Do YYYY"));
+      this.allDay?.setValue(!res?.allDay);
+      this.toggleAllDay();
+      this.allDay?.setValue(res?.allDay);
+      this.startTime?.setValue(res?.startTime);
+      this.endTime?.setValue(res?.endTime);
+      this.name?.setValue(res?.name);
+      this.details?.setValue(res?.details);
+    });
+  }
+
+  @Output() complete: EventEmitter<boolean> = new EventEmitter();
+
+  eventForm!: FormGroup;
+
+  constructor(private calendarService: CalendarService) {
+    this.eventForm = new FormGroup({
+      'date': new FormControl('', Validators.required),
+      'allDay': new FormControl(''),
+      'startTime': new FormControl('', Validators.required),
+      'endTime': new FormControl('', Validators.required),
+      'name': new FormControl('', Validators.required),
+      'details': new FormControl('', Validators.required)
+    })
+  }
+
+  ngOnDestroy(): void {
+    this.complete.complete();
+  }
+
+  get date() {
+    return this.eventForm.get('date');
+  }
+  get allDay() {
+    return this.eventForm.get('allDay');
+  }
+  get startTime() {
+    return this.eventForm.get('startTime');
+  }
+  get endTime() {
+    return this.eventForm.get('endTime');
+  }
+  get name() {
+    return this.eventForm.get('name');
+  }
+  get details() {
+    return this.eventForm.get('details');
+  }
+
+  dateSelected(value: moment.Moment) {
+    this.selectedDate = value;
+    this.date?.setValue(moment(value).format("dddd, MMMM Do YYYY"));
+  }
+
+  toggleAllDay() {
+    if (!this.allDay?.value) {
+      this.startTime?.clearValidators();
+      this.startTime?.updateValueAndValidity();
+      this.endTime?.clearValidators();
+      this.endTime?.updateValueAndValidity();
+    } else {
+      this.startTime?.addValidators(Validators.required);
+      this.startTime?.updateValueAndValidity();
+      this.endTime?.addValidators(Validators.required);
+      this.endTime?.updateValueAndValidity();
+    }
+  }
+
+  addEvent() {
+    if (this.eventForm.valid) {
+      let event: CalendarEvent = {
+        id: 10,
+        allDay: (this.allDay?.value ? true : false),
+        startTime: (this.allDay?.value ? null : this.startTime?.value),
+        endTime: (this.allDay?.value ? null : this.endTime?.value),
+        date: this.selectedDate,
+        name: this.name?.value,
+        details: this.details?.value
+      }
+      if (this.editing) {
+        event.id = this.id;
+        this.calendarService.editCalendarEvent(event).pipe(
+          take(1)
+        ).subscribe(_ => {
+          this.calendarService.refreshEvents();
+          this.complete.emit(true);
+          return;
```

Comment:
You don't need a return from the subscriptions

Reply:
Regarding the changes in the new commit:
The returns are removed, and the subscription itself functions the same, refreshing events and then emitting true. The form's job is complete after refreshing events, even if the subscribers of events or monthEvents haven't received the new/edited event yet.
I moved catchError to the subscriber's pipe rather than in the service's creation of the observable. This handles the emit false.